File: frontend/packages/renderer/src/api/context.ts
/**
 * #594 上下文组装 API（spec f6-context/gui-panel.md §2.1）：
 * POST /api/v1/context/assemble 预览上下文注入结果，前端据此渲染真实条目 + 勾选 override。
 */
import { apiFetch } from './client';

/** 上下文条目来源（7 字面量联合，与后端 Pydantic 契约对齐） */
export type ContextSourceType =
  | 'writing_requirements'
  | 'outline'
  | 'character_setting'
  | 'world_setting'
  | 'chapter_summary'
  | 'foreshadowing'
  | 'preference';

/** 单条上下文条目 */
export interface ContextItem {
  source: ContextSourceType;
  title: string;
  content: string;
  priority: number;
  metadata: Record<string, unknown>;
}

/** 组装块：条目 + 分层/预算信息 */
export interface ContextBlock {
  item: ContextItem;
  layer: string;
  token_count: number;
  compressed: boolean;
}

/** override 白名单（#1235）：显式数组 = 只注入命中项，空数组 = 删空（不注入）；缺省/null = 注入全部 */
export interface ContextOverride {
  character_ids: string[];
  foreshadowing_ids: string[];
  world_ids: string[];
}

/** 组装请求（writing_requirements 必填 min_length=1；max_tokens 可缺省或为 null） */
export interface AssembleContextRequest {
  project_id: string;
  chapter_id: string;
  model: string;
  writing_requirements: string;
  max_tokens?: number | null;
  override?: ContextOverride;
}

/** 被预算裁剪的条目 */
export interface ContextDropped {
  item: ContextItem;
  reason: string;
}

/** 组装结果 */
export interface ContextAssemblyResult {
  blocks: ContextBlock[];
  budget_tokens: number;
  total_tokens: number;
  model: string;
  dropped: ContextDropped[];
}

/** 预览上下文组装结果 */
export async function assembleContext(body: AssembleContextRequest): Promise<ContextAssemblyResult> {
  return apiFetch<ContextAssemblyResult>('/api/v1/context/assemble', { method: 'POST', body });
}

/** 世界观条目列表行（#704：选择注入搜索选择器数据源） */
export interface WorldSettingItem {
  id: string;
  name: string;
  category: string;
}

/** GET /api/v1/projects/{projectId}/world-settings——项目世界观条目列表 */
export async function listProjectWorldSettings(
  projectId: string,
): Promise<{ items: WorldSettingItem[]; total: number; offset: number; limit: number }> {
  return apiFetch(`/api/v1/projects/${projectId}/world-settings`);
}

/** 伏笔条目列表行（#704：选择注入搜索选择器数据源） */
export interface ForeshadowingItem {
  id: string;
  title: string;
  status: string;
  priority: number;
  location: string;
}

/** GET /api/v1/projects/{projectId}/foreshadowings——项目伏笔列表 */
export async function listProjectForeshadowings(
  projectId: string,
): Promise<{ items: ForeshadowingItem[]; total: number; offset: number; limit: number }> {
  return apiFetch(`/api/v1/projects/${projectId}/foreshadowings`);
}

/** 章节摘要 DTO（Issue #656）：与后端 context.py 契约对齐 */
export interface ChapterSummaryDto {
  summary: string;
  chapter_id: string;
}

/** 获取章节摘要（GET /api/v1/context/chapters/{chapterId}/summary，后端 ensure_summary 惰性生成） */
export async function getChapterSummary(chapterId: string): Promise<ChapterSummaryDto> {
  return apiFetch<ChapterSummaryDto>(`/api/v1/context/chapters/${chapterId}/summary`);
}

/** 强制刷新章节摘要（POST /api/v1/context/chapters/{chapterId}/summary/refresh，无 body） */
export async function refreshChapterSummary(chapterId: string): Promise<ChapterSummaryDto> {
  return apiFetch<ChapterSummaryDto>(`/api/v1/context/chapters/${chapterId}/summary/refresh`, {
    method: 'POST',
  });
}

/** #1349 章级注入记录（回执面）：本次生成**实际**注入了哪些设定条目 id */
export interface ChapterInjectionDto {
  chapter_id: string;
  /** 产生该明细的执行记录 id；无记录时为 null */
  execution_id: string | null;
  /** 三源 id 明细；无记录时为 null（前端据此回退 assemble 预览态） */
  injected_context: ContextOverride | null;
}

/** 章级注入记录回读（GET /api/v1/agent/chapters/{chapterId}/injections） */
export async function fetchChapterInjections(chapterId: string): Promise<ChapterInjectionDto> {
  return apiFetch<ChapterInjectionDto>(`/api/v1/agent/chapters/${chapterId}/injections`);
}

/** #1379 预选请求：进入空章时按本章大纲预挑相关条目（输入面与 assemble 同口径） */
export interface PreselectContextRequest {
  project_id: string;
  chapter_id: string;
  model: string;
  writing_requirements: string;
}

/**
 * #1379 预选结果（POST /api/v1/context/preselect）：
 * `mode="agent"` = LLM 按大纲预选出的三类 id 子集；
 * `mode="fallback"` = 回退全选（无大纲 / 预选失败），此时三类即全量候选。
 */
export interface ContextPreselectResult {
  character_ids: string[];
  world_ids: string[];
  foreshadowing_ids: string[];
  mode: 'agent' | 'fallback';
}

/** 按本章大纲预选相关条目（失败/不可用 → 调用方回退全选） */
export async function preselectContext(
  body: PreselectContextRequest,
): Promise<ContextPreselectResult> {
  return apiFetch<ContextPreselectResult>('/api/v1/context/preselect', { method: 'POST', body });
}

File: backend/src/inkflow/domain/models/context.py
"""上下文管理领域模型 — 分层 Token 预算、数据源类型、组装结果.

ContextLayer 定义三层上下文策略（protected / compressible / dynamic），
ContextSourceType 枚举七种数据源（F28 追加 preference），
ContextItem/ContextBlock/DroppedItem 为组装过程的
中间数据结构，ContextRequest 为 API 输入，ContextAssemblyResult 为最终产出。

依据: specs/f6-context/spec.md §3, ADR-010.
"""

from __future__ import annotations

import uuid
from dataclasses import dataclass, field
from enum import StrEnum
from typing import Any

from pydantic import BaseModel, Field, field_validator


class ContextLayer(StrEnum):
    """上下文分层 — 决定注入策略.

    - PROTECTED: 必须包含，不可压缩、不可裁剪
    - COMPRESSIBLE: 可摘要压缩，压缩后仍超预算才裁剪
    - DYNAMIC: 按预算择优选择，放不下直接裁剪
    """

    PROTECTED = "protected"
    COMPRESSIBLE = "compressible"
    DYNAMIC = "dynamic"


class ContextSourceType(StrEnum):
    """上下文数据源类型枚举.

    ┌─────────────────────────┬───────────────┬──────────────────────────┐
    │ 值                      │ 层            │ Phase 1 数据来源          │
    ├─────────────────────────┼───────────────┼──────────────────────────┤
    │ writing_requirements    │ protected     │ F3 调用时必传入参          │
    │ outline                 │ protected     │ project.config.extra      │
    │ character_setting       │ compressible  │ 空实现 (F8 Phase 2)       │
    │ world_setting           │ compressible  │ 空实现 (F9 Phase 2)       │
    │ chapter_summary         │ dynamic       │ LLM 生成 + 缓存表 (本模块) │
    │ foreshadowing           │ dynamic       │ 空实现 (F14 Phase 2)      │
    │ preference              │ protected     │ F28 已学偏好（memory_learning） │
    └─────────────────────────┴───────────────┴──────────────────────────┘
    """

    WRITING_REQUIREMENTS = "writing_requirements"
    OUTLINE = "outline"
    CHARACTER_SETTING = "character_setting"
    WORLD_SETTING = "world_setting"
    CHAPTER_SUMMARY = "chapter_summary"
    FORESHADOWING = "foreshadowing"
    PREFERENCE = "preference"


# ── 层级映射 ────────────────────────────────────────────────────────
SOURCE_LAYER: dict[ContextSourceType, ContextLayer] = {
    ContextSourceType.WRITING_REQUIREMENTS: ContextLayer.PROTECTED,
    ContextSourceType.OUTLINE: ContextLayer.PROTECTED,
    ContextSourceType.CHARACTER_SETTING: ContextLayer.COMPRESSIBLE,
    ContextSourceType.WORLD_SETTING: ContextLayer.COMPRESSIBLE,
    ContextSourceType.CHAPTER_SUMMARY: ContextLayer.DYNAMIC,
    ContextSourceType.FORESHADOWING: ContextLayer.DYNAMIC,
    ContextSourceType.PREFERENCE: ContextLayer.PROTECTED,
}


@dataclass
class ContextItem:
    """单一上下文条目 — 由数据源产出.

    Attributes:
        source: 来源类型.
        title: 注入时的分段标题，如「角色：林晚」「第 3 章摘要」.
        content: 内容文本.
        priority: 同层内优先级（大者先注入），默认 0.
        metadata: 扩展元数据，如 chapter_id / chapter_index / location.
    """

    source: ContextSourceType
    title: str
    content: str
    priority: int = 0
    metadata: dict[str, Any] = field(default_factory=dict)


@dataclass
class ContextBlock:
    """注入块 — 预算分配后的产物.

    Attributes:
        item: 原始上下文条目.
        layer: 所属层.
        token_count: 该 block 占用的 Token 数.
        compressed: 是否已被摘要压缩.
    """

    item: ContextItem
    layer: ContextLayer
    token_count: int
    compressed: bool = False


@dataclass
class DroppedItem:
    """被裁剪的条目及原因.

    Attributes:
        item: 原始条目.
        reason: over_budget / compression_insufficient / layer_cap / summary_failed.
    """

    item: ContextItem
    reason: str


@dataclass
class ChapterSummary:
    """前文摘要缓存 — 映射到 chapter_summaries 表.

    失效规则: chapter.updated_at > summary.updated_at 时缓存过期.
    """

    id: uuid.UUID
    chapter_id: uuid.UUID
    summary: str
    model: str
    created_at: str  # ISO 8601 UTC
    updated_at: str  # ISO 8601 UTC


class TokenBudgetConfig(BaseModel):
    """分层预算配置 — 存储在 project.config.extra["context"] 中.

    Attributes:
        max_ratio: 预算上限 = 模型窗口 × max_ratio (PRD: ≤ 80%).
        layer_ratio: 各层 cap 占比（需 ≤ 1.0，否则自动归一化）.
        summary_model: 摘要专用模型 (None = 用请求的 model).
        summary_max_chapters: dynamic 层最多注入的摘要数.
        compress_target_ratio: 压缩目标比例（压缩后 ≤ 原文 token × ratio）.
    """

    max_ratio: float = Field(default=0.8, ge=0.1, le=1.0)
    layer_ratio: dict[ContextLayer, float] = Field(
        default_factory=lambda: {
            ContextLayer.PROTECTED: 0.30,
            ContextLayer.COMPRESSIBLE: 0.40,
            ContextLayer.DYNAMIC: 0.30,
        }
    )
    summary_model: str | None = None
    summary_max_chapters: int = Field(default=10, ge=1, le=100)
    compress_target_ratio: float = Field(default=0.5, ge=0.1, le=1.0)

    @field_validator("layer_ratio")
    @classmethod
    def _normalize_layer_ratio(cls, v: dict[ContextLayer, float]) -> dict[ContextLayer, float]:
        total = sum(v.values())
        if total > 1.0:
            return {k: val / total for k, val in v.items()}
        return v


class ContextOverride(BaseModel):
    """上下文注入的显式勾选通道（v1.1 #593）.

    - character_ids 非空 → 只注入 metadata.character_id 命中的角色 item；空 → 注入全部
    - foreshadowing_ids 非空 → 只注入 metadata.foreshadowing_id 命中的伏笔 item；空 → 注入全部
    - world_ids 非空 → 只注入 metadata.world_setting_id 命中的世界观 item；空 → 注入全部
    - 只过滤 character_setting / foreshadowing / world_setting 三类来源，不影响 outline/summary 等
    """

    character_ids: list[uuid.UUID] = Field(default_factory=list)
    foreshadowing_ids: list[uuid.UUID] = Field(default_factory=list)
    world_ids: list[uuid.UUID] = Field(default_factory=list)


class ContextRequest(BaseModel):
    """上下文组装请求 — API 输入 / F3 调用参数.

    Attributes:
        project_id: 项目 ID.
        chapter_id: 目标章节 ID（正在写作的章节）.
        model: 目标模型名（provider/model_name 格式）.
        writing_requirements: 必填，写作要求 / 任务指令.
        max_tokens: 覆盖预算；None = 模型窗口 × max_ratio.
        override: 显式勾选通道（v1.1 #593）；None = 全部注入（默认行为）.
    """

    project_id: uuid.UUID
    chapter_id: uuid.UUID | None = None
    model: str
    writing_requirements: str = Field(..., min_length=1)
    max_tokens: int | None = None
    override: ContextOverride | None = None


class ContextAssemblyResult(BaseModel):
    """上下文组装结果 — F3 据此渲染最终 Prompt.

    Attributes:
        blocks: 按 protected → compressible → dynamic 有序排列.
        budget_tokens: 总预算 Token 数.
        total_tokens: 实际占用 Token 数.
        model: 目标模型.
        dropped: 被裁剪的条目清单位置.
    """

    blocks: list[ContextBlock]
    budget_tokens: int
    total_tokens: int
    model: str
    dropped: list[DroppedItem]
